Gems/MotionMatching/Code/Source/PoseDataJointVelocities.cpp

```diff
@@ -138,15 +140,70 @@ namespace EMotionFX::MotionMatching
 
     void PoseDataJointVelocities::CalculateVelocity(MotionInstance* motionInstance, size_t relativeToJointIndex)
     {
+        const size_t numJoints = m_velocities.size();
         SetRelativeToJointIndex(relativeToJointIndex);
         ActorInstance* actorInstance = motionInstance->GetActorInstance();
-        m_velocities.resize(actorInstance->GetNumNodes());
-        m_angularVelocities.resize(actorInstance->GetNumNodes());
-        for (size_t i = 0; i < m_velocities.size(); ++i)
+        m_velocities.resize(numJoints);
+        m_angularVelocities.resize(numJoints);
+
+        const float originalTime = motionInstance->GetCurrentTime();
+
+        // Prepare for sampling.
+        AnimGraphPosePool& posePool = GetEMotionFX().GetThreadData(actorInstance->GetThreadIndex())->GetPosePool();
+        AnimGraphPose* prevPose = posePool.RequestPose(actorInstance);
+        AnimGraphPose* currentPose = posePool.RequestPose(actorInstance);
+        Pose* bindPose = actorInstance->GetTransformData()->GetBindPose();
+
+        const size_t numSamples = 3;
```

Comment:
Maybe define this outside of the class? This means we always sample 3 frame ahead right?

Reply:
That's the size of the sliding window centered around the current time, so basically this means we're sampling one before and one after and one exactly at the requested time. Basically just to filter out the high frequency changes in velocity out and get a little more smoothed velocity.

I am still thinking about how we can make this code a little bit more sharable as this is also used for the base feature class, so yeah I am with you here! Will come in one of the following updates.